### PR TITLE
[Flink-3318][cep] Add support for quantifiers to CEP's pattern API

### PR DESCRIPTION
**Pull request by dawidwys**
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [x] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [x] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [x] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

### CODE REVIEW

**Comment by dawidwys** [comment]
@kl0u could you have a look at this PR?

**Comment by kl0u** [comment]
Hi @dawidwys , I will do this in the beginning of next week at the latest. 

One thing that I saw is that in the `Pattern` class you do not check for conflicting configurations e.g. oneToMany and zeroToMany. It would be nice to have that with a good error message.

**Comment by dawidwys** [comment]
Thanks @kl0u for the comment. I was wondering if it would be possible(and beneficial) to restrict it on the interface level and come up with a class hierachy like : 
```java
interface IPattern {
 /**
Getters for properties e.g.
 */ 

     String getName()
     Quantifier getQuantifier()
}

class Pattern implements IPattern {
     public Pattern<T, F> where(FilterFunction<F> newFilterFunction) 

     public Pattern<T, F> where(FilterFunction<F> newFilterFunction) 

     public QuantifiedPattern<T, F> zeroOrMore(final boolean eager) 
...
}

class QuantifiedPattern implements IPattern {
     private Pattern<T, F> wrappedPattern;

     public QuantifiedPattern<T, F> where(FilterFunction<F> newFilterFunction) 

     public String getName() {
        return wrappedPattern.getName();
     }

     public String getQuantifier() {
        return wrappedPattern.getQuantifier();
     }
/* no quantifier specific methods */
}
```

You can see a working example for java-cep in my fork: https://github.com/dawidwys/flink/tree/cepQuantifierAPI

Not sure if the hierarchy is not too complex, and simple checks with exceptions would be better.
We would also need to duplicate this hierarchy in the scala part.
@kl0u What do you think?

**Comment by kl0u** [comment]
Hi @dawidwys , I will start reviewing your PR now. 

I would say that for now it makes sense to have it with exceptions and not change the `Pattern` hierarchy. The reason is:
1) not sure what the above split will imply to backwards compatibility
2) there are some additional features that will be nice to have, and they all affect also how patterns are defined. A refactoring of the `Pattern` hierarchy would make more sense as soon as we have everything in place, and we can factor in this knowledge when designing.

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java [hunk 1]
```diff
@@ -71,19 +70,16 @@
 	private final NonDuplicatingTypeSerializer<T> nonDuplicatingTypeSerializer;
 
 	// Buffer used to store the matched events
-	private final SharedBuffer<State<T>, T> sharedBuffer;
+	private final SharedBuffer<String, T> sharedBuffer;
 
 	// Set of all NFA states
-	private final Set<State<T>> states;
```
Comment (kl0u) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java]: Why `List` and not `Set`?

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java [hunk 2]
```diff
@@ -209,33 +204,66 @@ public boolean equals(Object obj) {
...
 			return false;
 		}
 	}
 
 	@Override
 	public int hashCode() {
-		return Objects.hash(nonDuplicatingTypeSerializer, sharedBuffer, states, windowTime, startEventCounter);
+		return Objects.hash(nonDuplicatingTypeSerializer, sharedBuffer, states, windowTime);
+	}
+
```
Comment (kl0u) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java]: Why is this method is needed? Is the name comparison enough? And if it is needed indeed, I think it should move to the `State` class with a comment on why it is needed.

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java [hunk 3]
```diff
@@ -209,33 +204,66 @@ public boolean equals(Object obj) {
...
+		}
+		int getTotalTakeBranches() {
+			return totalTakeBranches;
 		}
-	};
+
+		List<StateTransition<T>> getEdges() {
+			return edges;
+		}
+
```
Comment (kl0u) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java]: Why is this method needed? The two end-points of the IGNORE edge, won't they be the same state?

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java [hunk 4]
```diff
@@ -247,27 +275,148 @@ public int compare(final StateTransition<T> o1, final StateTransition<T> o2) {
...
+		final long timestamp) {
+		final ArrayList<ComputationState<T>> resultingComputationStates = new ArrayList<>();
+
+		final OutgoingEdges<T> outgoingEdges = createDecisionGraph(computationState, event);
+
+		// Create the computing version based on the previously computed edges
+		// We need to defer the creation of computation states until we know how many edges start
+		// at this computation state so that we can assign proper version
+		final List<StateTransition<T>> edges = outgoingEdges.getEdges();
+		Integer takeBranchesToVisit = Math.max(0, outgoingEdges.getTotalTakeBranches() - 1);
```
Comment (kl0u) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java]: Unbox these two `Integer`s to `int`s.

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java [hunk 5]
```diff
@@ -247,27 +275,148 @@ public int compare(final StateTransition<T> o1, final StateTransition<T> o2) {
...
-			final ComputationState<T> computationState,
-			final T event,
-			final long timestamp) {
-		Stack<State<T>> states = new Stack<>();
-		ArrayList<ComputationState<T>> resultingComputationStates = new ArrayList<>();
-		State<T> state = computationState.getState();
+		final ComputationState<T> computationState,
+		final T event,
+		final long timestamp) {
+		final ArrayList<ComputationState<T>> resultingComputationStates = new ArrayList<>();
```
Comment (kl0u) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java]: Move the `resultingComputationStates` just before the `for (StateTransition<T> edge : edges) {...}`.

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java [hunk 6]
```diff
@@ -247,27 +275,148 @@ public int compare(final StateTransition<T> o1, final StateTransition<T> o2) {
...
+		for (StateTransition<T> edge : edges) {
+			switch (edge.getAction()) {
+				case IGNORE: {
+					if (!computationState.isStartState()) {
+						final DeweyNumber version;
+						if (!isEquivalentState(edge.getTargetState(), computationState.getState())) {
+							version = computationState.getVersion().increase(ignoreBranchesToVisit).addStage();
+							ignoreBranchesToVisit--;
+						} else {
+							final int toIncrease = calculateIncreasingSelfState(outgoingEdges.getTotalIgnoreBranches(),
```
Comment (kl0u) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java]: Move both the arguments to different lines.

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java [hunk 7]
```diff
@@ -247,27 +275,148 @@ public int compare(final StateTransition<T> o1, final StateTransition<T> o2) {
...
+				computationState.getState().getName(),
+				computationState.getEvent(),
+				computationState.getTimestamp());
+		}
+
+		return resultingComputationStates;
+	}
+
+	private int calculateIncreasingSelfState(int ignoreBranches, int takeBranches) {
+		if (takeBranches == 0 && ignoreBranches == 0) {
```
Comment (kl0u) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java]: This can become: `return takeBranches == 0 && ignoreBranches == 0 ? 0 : ignoreBranches + 1;`

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java [hunk 8]
```diff
@@ -247,27 +275,148 @@ public int compare(final StateTransition<T> o1, final StateTransition<T> o2) {
...
+			return 0;
+		}
+
+		return ignoreBranches + 1;
+	}
+
```
Comment (kl0u) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java]: Move arguments in the same line

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java [hunk 9]
```diff
@@ -247,27 +275,148 @@ public int compare(final StateTransition<T> o1, final StateTransition<T> o2) {
...
+	private ComputationState<T> createStartState(final ComputationState<T> computationState,
+		final int totalBranches) {
+		final DeweyNumber startVersion = computationState.getVersion().increase(totalBranches);
 
-		boolean branched = false;
+		return ComputationState.createStartState(
```
Comment (kl0u) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java]: Move arguments in the same line

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/SharedBuffer.java [hunk 10]
```diff
@@ -217,23 +217,22 @@ public void prune(long pruningTimestamp) {
...
 					while(!currentPath.isEmpty()) {
-						SharedBufferEntry<K, V> currentEntry = currentPath.pop();
+						final SharedBufferEntry<K, V> currentEntry = currentPath.pop();
 
 						completePath.put(currentEntry.getKey(), currentEntry.getValueTime().getValue());
 					}
 
 					result.add(completePath);
 				} else {
-					SharedBufferEntry<K, V> currentEntry = extractionState.getEntry();
```
Comment (kl0u) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/SharedBuffer.java]: The `SharedBufferEntry<K, V> currentEntry = extractionState.getEntry();` can be moved outside the for-loop right?

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/StateTransition.java [hunk 11]
```diff
@@ -55,6 +65,7 @@ public boolean equals(Object obj) {
 			StateTransition<T> other = (StateTransition<T>) obj;
```
Comment (kl0u) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/StateTransition.java]: You have updated the `equals`, you should also update the `hashCode`  and the `toString` accordingly.

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/pattern/InvalidPatternException.java [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.flink.cep.pattern;
+
+/**
+ * Thrown when the given {@link Pattern} was constructed illegaly e.g when
+ * more than quantifier provided.
+ */
+class InvalidPatternException extends RuntimeException {
```
Comment (kl0u) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/pattern/InvalidPatternException.java]: Instead of creating a new `Exception` I suggest we should use the existing `MalformedPatternException`. You can move it to this package, as it is used more broadly now.

File: flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/compiler/NFACompilerTest.java [hunk 13]
```diff
@@ -18,9 +18,13 @@
 
 package org.apache.flink.cep.nfa.compiler;
```
Comment (kl0u) [inline flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/compiler/NFACompilerTest.java]: Unused imports.

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/pattern/Quantifier.java [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.flink.cep.pattern;
+
+public enum Quantifier {
```
Comment (kl0u) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/pattern/Quantifier.java]: Here you can add some flags, like `isLooping`, `isAtLeastOnce`, etc and set them accordingly in the constructors. This will allow you to move all the `is*` methods like `isLooping()`, `isAtLeastOnce` from the `NFACompiler` here.

File: flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/NFAITCase.java [hunk 15]
```diff
@@ -272,37 +372,42 @@ public void testBranchingPattern() {
...
-		})
-			.followedBy("middle-first").subtype(SubEvent.class).where(new FilterFunction<SubEvent>() {
+		}).followedBy("middle-first").subtype(SubEvent.class)
+			.where(new FilterFunction<SubEvent>() {
 				private static final long serialVersionUID = 6215754202506583964L;
 
 				@Override
 				public boolean filter(SubEvent value) throws Exception {
-					return value.getVolume() > 5.0;
+					return value.getVolume() >
```
Comment (kl0u) [inline flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/NFAITCase.java]: this can be as before, in one line. In general we try to minimize unnecessary changes ;).

File: flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/NFAITCase.java [hunk 16]
```diff
@@ -272,37 +372,42 @@ public void testBranchingPattern() {
...
+						.equals("end");
 				}
 			});
 
 		NFA<Event> nfa = NFACompiler.compile(pattern, Event.createTypeSerializer(), false);
 
 		List<Map<String, Event>> resultingPatterns = new ArrayList<>();
 
-		for (StreamRecord<Event> inputEvent: inputEvents) {
+		for (StreamRecord<Event> inputEvent : inputEvents) {
```
Comment (kl0u) [inline flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/NFAITCase.java]: this can be as before, in one line.

File: flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/NFAITCase.java [hunk 17]
```diff
@@ -272,37 +372,42 @@ public void testBranchingPattern() {
...
+							"next-one");
 				}
 			})
 			.followedBy("end").where(new FilterFunction<Event>() {
 				private static final long serialVersionUID = 7056763917392056548L;
 
 				@Override
 				public boolean filter(Event value) throws Exception {
-					return value.getName().equals("end");
+					return value.getName()
```
Comment (kl0u) [inline flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/NFAITCase.java]: this can be as before, in one line.

File: flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/NFAITCase.java [hunk 18]
```diff
@@ -272,37 +372,42 @@ public void testBranchingPattern() {
 			public boolean filter(Event value) throws Exception {
 				return value.getName().equals("start");
 			}
-		})
-			.followedBy("middle-first").subtype(SubEvent.class).where(new FilterFunction<SubEvent>() {
+		}).followedBy("middle-first").subtype(SubEvent.class)
```
Comment (kl0u) [inline flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/NFAITCase.java]: this can be as before, in one line.

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java (hunk 5 above)
Comment (dawidwys) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java]: Done

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java (hunk 4 above)
Comment (dawidwys) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java]: Done

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java (hunk 6 above)
Comment (dawidwys) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java]: Done

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java (hunk 7 above)
Comment (dawidwys) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java]: Done

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java (hunk 8 above)
Comment (dawidwys) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java]: Done

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java (hunk 9 above)
Comment (dawidwys) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java]: Done

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/SharedBuffer.java (hunk 10 above)
Comment (dawidwys) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/SharedBuffer.java]: If you mean outside the `if-else` then right. I've done that.

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/StateTransition.java (hunk 11 above)
Comment (dawidwys) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/StateTransition.java]: Done

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/pattern/InvalidPatternException.java (hunk 12 above)
Comment (dawidwys) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/pattern/InvalidPatternException.java]: Done.

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/pattern/Quantifier.java (hunk 14 above)
Comment (dawidwys) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/pattern/Quantifier.java]: Done

File: flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/NFAITCase.java (hunk 15 above)
Comment (dawidwys) [inline flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/NFAITCase.java]: I reverted the whole `testSimplePatternNFA`. Probably I've run formatter on whole file.

File: flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/compiler/NFACompilerTest.java (hunk 13 above)
Comment (dawidwys) [inline flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/compiler/NFACompilerTest.java]: Done

File: flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/NFAITCase.java (hunk 17 above)
Comment (dawidwys) [inline flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/NFAITCase.java]: Reverted

File: flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/NFAITCase.java (hunk 18 above)
Comment (dawidwys) [inline flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/NFAITCase.java]: Reverted

File: flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/NFAITCase.java (hunk 16 above)
Comment (dawidwys) [inline flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/NFAITCase.java]: Reverted

**Comment by dawidwys** [comment]
Thanks @kl0u for your comments, I've changed all code style specific. 

Some comments were addressing a behaviour that one `Pattern` can be translated into multiple "equivalent" `States` with the same name.

**Comment by kl0u** [comment]
Thanks @dawidwys I will have a look now.

**Comment by kl0u** [comment]
Hi @dawidwys , the following produces no output, which should not be the case. The same holds if you change the pattern to `oneOrMore`.

```
public void testZeroOrMore() {
		List<StreamRecord<Event>> inputEvents = new ArrayList<>();

		Event startEvent = new Event(40, "c", 1.0);
		Event middleEvent1 = new Event(41, "a", 2.0);
		Event middleEvent2 = new Event(42, "a", 3.0);
		Event middleEvent3 = new Event(43, "a", 4.0);
		Event end1 = new Event(44, "b", 5.0);
		Event end2 = new Event(45, "d", 6.0);
		Event end3 = new Event(46, "d", 7.0);
		Event end4 = new Event(47, "e", 8.0);

		inputEvents.add(new StreamRecord<>(startEvent, 1));
		inputEvents.add(new StreamRecord<>(middleEvent1, 3));
		inputEvents.add(new StreamRecord<>(middleEvent2, 4));
		inputEvents.add(new StreamRecord<>(middleEvent3, 5));
		inputEvents.add(new StreamRecord<>(end1, 6));
		inputEvents.add(new StreamRecord<>(end2, 7));
		inputEvents.add(new StreamRecord<>(end3, 8));
		inputEvents.add(new StreamRecord<>(end4, 9));

		Pattern<Event, ?> pattern = Pattern.<Event>begin("start").where(new FilterFunction<Event>() {
			private static final long serialVersionUID = 5726188262756267490L;

			@Override
			public boolean filter(Event value) throws Exception {
				return value.getName().equals("a");
			}
		}).zeroOrMore();

		NFA<Event> nfa = NFACompiler.compile(pattern, Event.createTypeSerializer(), false);

		Set<Set<Event>> resultingPatterns = new HashSet<>();
		List<Collection<Event>> allPatterns = new ArrayList<>();

		for (StreamRecord<Event> inputEvent : inputEvents) {
			Collection<Map<String, Event>> patterns = nfa.process(
					inputEvent.getValue(),
					inputEvent.getTimestamp()).f0;

			for (Map<String, Event> foundPattern : patterns) {
				System.out.println(foundPattern);
			}
		}
	}
```

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/compiler/NFACompiler.java [hunk 19]
```diff
@@ -74,88 +76,233 @@
...
-				succeedingState = currentState;
-				currentPattern = currentPattern.getPrevious();
+		private State<T> createMiddleStates(final State<T> sinkState) {
 
-				if (!patternNames.add(currentPattern.getName())) {
-					throw new MalformedPatternException("Duplicate pattern name: " + currentPattern.getName() + ". " +
-						"Pattern names must be unique.");
+			State<T> lastSink = sinkState;
+			while (currentPattern.getPrevious() != null) {
+				State<T> sourceState;
```
Comment (kl0u) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/compiler/NFACompiler.java]: You do not need to declare it here and set it 3 lines below, you can just create the variable later when you set it:
`State<T> sourceState = new State<>(currentPattern.getName(), State.StateType.Normal);`

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/compiler/NFACompiler.java [hunk 20]
```diff
@@ -74,88 +76,233 @@
...
+				sourceState = new State<>(currentPattern.getName(), State.StateType.Normal);
+				states.add(sourceState);
+				convertToSingletonState(sinkState, sourceState);
+			}
+			return sourceState;
+		}
+
+		@SuppressWarnings("unchecked")
+		private void convertToSingletonState(
+			final State<T> sinkState,
```
Comment (kl0u) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/compiler/NFACompiler.java]: From what I understand, the `ignoredState` is needed only in the case where we have a `FollowedByPattern`. In this case the code could become:

```
private void convertToSingletonState(
			final State<T> sinkState,
			final State<T> sourceState) {

			final FilterFunction<T> currentFilterFunction = 
                                         (FilterFunction<T>) currentPattern.getFilterFunction();
			final FilterFunction<T> trueFunction = FilterFunctions.trueFunction();
			sourceState.addTake(sinkState, currentFilterFunction);

			if (currentPattern.getQuantifier() == Quantifier.OPTIONAL) {
				sourceState.addProceed(sinkState, trueFunction);
			}

			if (currentPattern instanceof FollowedByPattern) {
				final State<T> ignoreState;
				if (currentPattern.getQuantifier() == Quantifier.OPTIONAL) {
					ignoreState = new State<>(currentPattern.getName(), State.StateType.Normal);
					ignoreState.addTake(sinkState, currentFilterFunction);
					states.add(ignoreState);
				} else {
					ignoreState = sourceState;
				}
				sourceState.addIgnore(ignoreState, trueFunction);
			}
		}
```

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/compiler/NFACompiler.java [hunk 21]
```diff
@@ -74,88 +76,233 @@
...
-			// add the beginning state
-			final State<T> beginningState;
+			states.add(beginningState);
+			usedNames.add(beginningState.getName());
+
+			return beginningState;
+		}
+
+		private State<T> convertToTimesState(State<T> sinkState, State<T> sourceState, int times) {
+			convertToSingletonState(sinkState, sourceState);
```
Comment (kl0u) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/compiler/NFACompiler.java]: Here I have 3 comments: 
1) you can create local variables instead of reusing `sourceState` and `sinkState`.
2) you can add a check like `Preconditions.checkArgument(currentPattern.getQuantifier() == Quantifier.TIMES)`
3) currently we cannot have `OPTIONAL` and `TIMES` but the code seems to support that. Am I correct? In this case we could also have in the `Pattern` a method called `atMost(int times)` which will allow the user to specify such a pattern. What do you think?

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/compiler/NFACompiler.java [hunk 22]
```diff
@@ -74,88 +76,233 @@
...
+			if (currentPattern instanceof FollowedByPattern) {
+				sourceState.addIgnore(ignoreState, trueFunction);
+			}
+		}
+
+		@SuppressWarnings("unchecked")
+		private State<T> createFirstMandatoryStateOfLoop(
+			final State<T> sinkState,
+			final State.StateType stateType) {
+
```
Comment (kl0u) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/compiler/NFACompiler.java]: You can add a check: 
`Preconditions.checkArgument(currentPattern.getQuantifier().isAtLeastOne());`

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/compiler/NFACompiler.java [hunk 23]
```diff
@@ -74,88 +76,233 @@
...
+			final FilterFunction<T> currentFilterFunction = (FilterFunction<T>) currentPattern.getFilterFunction();
+			final State<T> firstState = new State<>(currentPattern.getName(), stateType);
+
+			firstState.addTake(sinkState, currentFilterFunction);
+			if (currentPattern instanceof FollowedByPattern) {
+				if (currentPattern.getQuantifier() == Quantifier.ONE_OR_MORE_COMBINATIONS) {
```
Comment (kl0u) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/compiler/NFACompiler.java]: In the `Quantifier` you can add a flag like `isCombinations` for the `ONE_OR_MORE_COMBINATIONS` and the `ZERO_OR_MORE_COMBINATIONS` and check it here. Given that you already check for `atLeastOne` in the check I proposed above, this will be sufficient.

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/compiler/NFACompiler.java [hunk 24]
```diff
@@ -74,88 +76,233 @@
...
-				currentState,
-				(FilterFunction<T>) currentPattern.getFilterFunction()
-			));
+		@SuppressWarnings("unchecked")
+		private void convertToLooping(
+			final State<T> sinkState,
+			final State<T> sourceState) {
 
-			return new NFAFactoryImpl<T>(inputTypeSerializer, windowTime, new HashSet<>(states.values()), timeoutHandling);
+			final FilterFunction<T> filterFunction = (FilterFunction<T>) currentPattern.getFilterFunction();
```
Comment (kl0u) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/compiler/NFACompiler.java]: You can add a check: `Preconditions.checkArgument(currentPattern.getQuantifier().isLooping());`

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/compiler/NFACompiler.java [hunk 25]
```diff
@@ -74,88 +76,233 @@
...
+			sourceState.addProceed(sinkState, trueFunction);
+			sourceState.addTake(filterFunction);
+			if (currentPattern instanceof FollowedByPattern) {
+				final State<T> ignoreState = new State<>(
+					currentPattern.getName(),
+					State.StateType.Normal);
+
+
+				final FilterFunction<T> ignoreCondition;
+				if (currentPattern.getQuantifier() == Quantifier.ZERO_OR_MORE_COMBINATIONS ||
```
Comment (kl0u) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/compiler/NFACompiler.java]: If you add the flag I proposed above in the `Quantifier` class, this check will become clearer I think (`currentPattern.getQuantifier().isCombinations()`) or something like this.

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/ComputationState.java [hunk 26]
```diff
@@ -74,11 +80,35 @@ public long getStartTimestamp() {
...
+	public static <T> ComputationState<T> createStartState(final State<T> state) {
+		Preconditions.checkArgument(state.isStart());
+		return new ComputationState<>(state, null, null, -1L, new DeweyNumber(1), -1L);
+	}
+
+	public static <T> ComputationState<T> createStartState(final State<T> state, final DeweyNumber version) {
+		Preconditions.checkArgument(state.isStart());
+		return new ComputationState<>(state, null, null, -1L, version, -1L);
+	}
+
```
Comment (kl0u) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/ComputationState.java]: Indent the arguments to separate them from the body of the method. This will make the code easier to read.

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/compiler/NFACompiler.java (hunk 22 above)
Comment (dawidwys) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/compiler/NFACompiler.java]: I think it is an internal method in internal class. Not sure if we should be so defensive. If the check fails it means some error in implemented algorithm not in provided arguments.

Comment (kl0u) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/compiler/NFACompiler.java]: I agree that it is defensive. The only reason is that:
1) why not being defensive on these things? :P
2) I believe the code is more readable if the assumptions are clear when you start reading a method (although also the name can play the same role). If you see this check, then you know what is happening.

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/compiler/NFACompiler.java (hunk 23 above)
Comment (dawidwys) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/compiler/NFACompiler.java]: I changed the `Quantifier`. How do you like it now?

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/compiler/NFACompiler.java (hunk 21 above)
Comment (dawidwys) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/compiler/NFACompiler.java]: Ad.3 There is one problem with the atMost. Consider atMost(2).optional().where(a). It will be translated into states: 
`q1? q2?` . For a sequence `a1 a2`. It will generate matches `a1; a1; a2; a2; a1 a2`. Notice duplicated `a1` and `a2`. It is because a single event can match either on `q1` or at `q2`.

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/compiler/NFACompiler.java (hunk 20 above)
Comment (dawidwys) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/compiler/NFACompiler.java]: I've done it the other way round as to handle optional in one place. But I think your version is easier to analyze.

**Comment by dawidwys** [comment]
Hi, I've addressed the style comments. Still have some doubts about the `Preconditions`. I will try to add more comments on the `NFACompiler` with the implementation that will handle the failing test case. So I believe it will be okay to wait with further reviewing until then.

**Comment by kl0u** [comment]
Sounds good! Thanks @dawidwys

**Comment by dawidwys** [comment]
I've updated the PR with a solution for the previously failing test. Also tried to add some more comments. @kl0u I think it is ready for another round of review :)

**Comment by kl0u** [comment]
Hi @dawidwys , I have 2 comments for this PR:
1) in the `findFinalStateAfterProceed()` in the `NFA` you should also check the filter condition as you traverse the `PROCEED` edges towards the end.
2) In this PR the computation states are consuming, and the `endState` is a "dummy" state.
Given this, in the `extractPatternMatches()` in the NFA, you do a 
```
sharedBuffer.extractPatterns(
            computationState.getPreviousState().getName(),
            computationState.getEvent(),
            computationState.getTimestamp(),
            computationState.getVersion());
```
to extract the matched patterns.

This will work for patterns that terminate properly, as the `endState` now is a "dummy" state and you do not want to present it. But if I understand correctly, for timed out patterns, this means that the last state will not be presented. Is this correct?

**Comment by dawidwys** [comment]
Ad. 2 It will also work for not terminated patterns, as the `computationState` has not yet consumed any events (consuming state), so the first state that has any elements in `sharedBuffer` is the previous one.

**Comment by kl0u** [comment]
Merged this! Thanks for the work @dawidwys. 
Can you close this PR and the JIRA?

**Comment by dawidwys** [comment]
Great! Thanks for your help @kl0u !

**Comment by eliaslevy** [comment]
Am I missing something or is there no way to get access to access to multiple events matched by these new quantifiers?  The `PatternSelectFunction.select` takes an argument of `Map<String,IN>` and not  `Map<String,IN[]>`.  

Which is the one being passed to `select` in the `Map`?  The last matched?  The first matched?

**Comment by dawidwys** [comment]
Right now if the pattern name is "a" the events will be returned with keys: "a[0]", "a[1]" and so on, but agree it may be counterintuitive. Please file a JIRA for it.

**Comment by kl0u** [comment]
Hi @eliaslevy , @dawidwys is right, in fact if the state's name is "a" and it has 2 matching events, there will be two returned keys "a_0" and "a_1". This is definitely counterintuitive and the reason it is still there is for backwards compatibility with previous versions. 

I think this should be addressed but I would also suggest to also start a discussion in the mailing list to see if people are ok with this. Unfortunately we are not aware if anyone is using the `CEP` library at the moment.

**Comment by eliaslevy** [comment]
I created [FLINK-6419](https://issues.apache.org/jira/browse/FLINK-6419).
